### PR TITLE
Accessy client

### PR DESCRIPTION
**Pull request by emanuelen5**
Client for performing essential tasks for handling accessy members (ORG and group memberships)

### CODE REVIEW

File: api/src/multiaccessy/accessy.py [hunk 1]
```diff
@@ -1,32 +1,36 @@
...
         raise RuntimeError(msg)
 
 
 @dataclass
 class AccessySession:
-    session_token: uuid
+    session_token: UUID
 
     @classmethod
     def create_session(cls, client_id: str, client_secret: str) -> "AccessySession":
```
Comment (andersroos) [inline api/src/multiaccessy/accessy.py:36]: Do not check the parameters like this. Makes it hard to make something string like and use that.

File: api/src/multiaccessy/accessy.py [hunk 2]
```diff
@@ -42,107 +46,215 @@ def create_session(cls, client_id: str, client_secret: str) -> "AccessySession":
...
+            users_in_org = self._get_users_org()
+
+        for item in users_in_org:
+            if item.get("msisdn", None) == phone_number:
+                user_id = item["id"]
+                return self._user_ids_to_accessy_members([user_id])[0]
+        else:
+            return None
+    
+    #################
```
Comment (andersroos) [inline api/src/multiaccessy/accessy.py:176]: Maybe easier to read if public methods are at the top.

File: api/src/multiaccessy/accessy.py [hunk 3]
```diff
@@ -42,107 +43,215 @@ def create_session(cls, client_id: str, client_secret: str) -> "AccessySession":
...
+        elif len(data) == 0:
+            raise ValueError("The API key does not have a corresponding organization membership")
+
+        return data[0]["organizationId"]
+
+    def _get_json(self, url: str, msg: str = None) -> dict:
+        """ Convenience method for getting data from a JSON endpoint that is not paginated """
+        response = requests.get(ACCESSY_URL + url,
+                                headers={"Authorization": f"Bearer {self.session_token}"})
+        check_response_error(response, f"{msg}")
```
Comment (andersroos) [inline api/src/multiaccessy/accessy.py:67]: f"{msg} == msg (iaf om msg är en str)

File: api/src/multiaccessy/accessy.py [hunk 4]
```diff
@@ -42,107 +43,215 @@ def create_session(cls, client_id: str, client_secret: str) -> "AccessySession":
...
+        return self._get_users_in_access_group(ACCESSY_SPECIAL_ACCESS_GROUP)
+    
+    def _user_ids_to_accessy_members(self, user_ids: list[UUID]) -> list["AccessyMember"]:
+        """ Convert a list of User ID:s to AccessyMembers """
+
+        APPLICATION_PHONE_NUMBER = object()  # Sentinel phone number for applications
+        def fill_user_details(user: "AccessyMember", user_id: str):
+            data = self._get_json(f"/org/admin/user/{user_id}")
+
+            # API keys do not have phone numbers
```
Comment (andersroos) [inline api/src/multiaccessy/accessy.py:129]: Improve comment about why this is needed?

File: api/src/multiaccessy/accessy.py [hunk 5]
```diff
@@ -42,107 +43,215 @@ def create_session(cls, client_id: str, client_secret: str) -> "AccessySession":
...
+            user.last_name = data.get("lastName", "")
+
+        def fill_membership_id(user: "AccessyMember", user_id: str):
+            data = self._get_json(f"/asset/admin/user/{user_id}/organization/{self.organization_id}/membership")
+            user.membership_id = data["id"]
+        
+        threads = []
+        accessy_members = []
+        for uid in user_ids:
+            accessy_member = AccessyMember(uid, None)
```
Comment (andersroos) [inline api/src/multiaccessy/accessy.py:148]: None is probably the default so this None can be removed.

File: api/src/multiaccessy/accessy.py [hunk 6]
```diff
@@ -42,107 +43,215 @@ def create_session(cls, client_id: str, client_secret: str) -> "AccessySession":
...
+
+    def _get_users_special(self) -> list[dict]:
+        """ Get all user ID:s with special access """
+        return self._get_users_in_access_group(ACCESSY_SPECIAL_ACCESS_GROUP)
+    
+    def _user_ids_to_accessy_members(self, user_ids: list[UUID]) -> list["AccessyMember"]:
+        """ Convert a list of User ID:s to AccessyMembers """
+
+        APPLICATION_PHONE_NUMBER = object()  # Sentinel phone number for applications
+        def fill_user_details(user: "AccessyMember", user_id: str):
```
Comment (andersroos) [inline api/src/multiaccessy/accessy.py:126]: Räcker väl med user som arg eftersom den har id i sig?

File: api/src/multiaccessy/accessy.py [hunk 7]
```diff
@@ -42,107 +43,215 @@ def create_session(cls, client_id: str, client_secret: str) -> "AccessySession":
...
+                return
+
+            try:
+                user.phone_number = data["msisdn"]
+            except KeyError:
+                logger.error(f"User {user.user_id} does not have a phone number. {data=}")
+            user.first_name = data.get("firstName", "")
+            user.last_name = data.get("lastName", "")
+
+        def fill_membership_id(user: "AccessyMember", user_id: str):
```
Comment (andersroos) [inline api/src/multiaccessy/accessy.py:141]: Samma här user har väl alltid id i sig?

File: api/src/multiaccessy/accessy.py [hunk 8]
```diff
@@ -42,107 +43,215 @@ def create_session(cls, client_id: str, client_secret: str) -> "AccessySession":
...
+                user_id = item["id"]
+                return self._user_ids_to_accessy_members([user_id])[0]
+        else:
+            return None
+    
+    #################
+    # Public methods
+    #################
+
+    def get_all_members(self) -> tuple[list["AccessyMember"], list["AccessyMember"], list["AccessyMember"]]:
```
Comment (andersroos) [inline api/src/multiaccessy/accessy.py:180]: Why not return a dataclass with the three lists? Or a named tuple?

**Review by andersroos — APPROVED** [review]

File: api/src/multiaccessy/accessy.py (hunk 1 above)
Comment (emanuelen5) [reply api/src/multiaccessy/accessy.py:36]: Removed check

File: api/src/multiaccessy/accessy.py (hunk 2 above)
Comment (emanuelen5) [reply api/src/multiaccessy/accessy.py]: Moved public to top

File: api/src/multiaccessy/accessy.py (hunk 8 above)
Comment (emanuelen5) [reply api/src/multiaccessy/accessy.py]: Created dataclass for the three lists

File: api/src/multiaccessy/accessy.py (hunk 4 above)
Comment (emanuelen5) [reply api/src/multiaccessy/accessy.py]: True, updated the comment :)

**Pull request merged**